core/org.openjdk.jmc.flightrecorder.rules/src/main/java/org/openjdk/jmc/flightrecorder/rules/util/RulesToolkit.java

```diff
@@ -1208,12 +1213,57 @@ public static String toRatioPercentString(IQuantity antecedent, IQuantity conseq
 		if (nThreads < 1) {
 			nThreads = Runtime.getRuntime().availableProcessors();
 		}
+		ResultProvider resultProvider = new ResultProvider();
 		Map<IRule, Future<IResult>> resultFutures = new HashMap<>();
 		Queue<RunnableFuture<IResult>> futureQueue = new ConcurrentLinkedQueue<>();
+		List<IRule> unavailableRules = new ArrayList<>();
+		List<IRule> rulesWithDependencies = new ArrayList<>();
 		for (IRule rule : rules) {
-			RunnableFuture<IResult> resultFuture = rule.createEvaluation(items, preferences, null);
-			resultFutures.put(rule, resultFuture);
-			futureQueue.add(resultFuture);
+			if (matchesEventAvailabilityMap(items, rule.getRequiredEvents())) {
+				if (hasDependency(rule)) {
+					rulesWithDependencies.add(rule);
+				} else {
+					RunnableFuture<IResult> resultFuture = rule.createEvaluation(items, preferences, resultProvider);
+					resultFutures.put(rule, resultFuture);
+					futureQueue.add(resultFuture);
+				}
+			} else {
+				unavailableRules.add(rule);
+				resultFutures.put(rule, evaluationErrorResult(rule, preferences));
+			}
+		}
+		for (IRule rule : rulesWithDependencies) {
+			String dependencyName = getRuleDependencyName(rule);
+			boolean shouldEvaluate = true;
+			for (IRule unavailableRule : unavailableRules) {
+				if (dependencyName.equals(unavailableRule.getId())) {
+					shouldEvaluate = false;
+					resultFutures.put(rule, evaluationErrorResult(rule, preferences));
+					break;
+				}
+			}
```

Comment:
This is not enough to determine whether or not a rule should be evaluated. The `@dependsOn` annotation specifies both a class and an optional minimum severity that the prior rule must return for the current rule to evaluate. If we disregard this we risk running into NPEs for later rules as they attempt to access results that might not be present in the RulesProvider.

Reply:
Right, thanks for pointing that out. I'll put a severity check into there similar to how it's done in RuleManager: https://github.com/openjdk/jmc/blob/master/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/RuleManager.java#L114